polkadot/xcm/pallet-xnft/src/lib.rs

```diff
@@ -0,0 +1,287 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![recursion_limit = "256"]
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{
+	pallet_prelude::*,
+	traits::{
+		tokens::asset_ops::{
+			common_asset_kinds::Instance,
+			common_strategies::{DeriveAndReportId, Owned, PredefinedId},
+			AssetDefinition, Create,
+		},
+		Incrementable,
+	},
+};
+use sp_runtime::DispatchResult;
+use sp_std::prelude::*;
+use xcm::latest::prelude::*;
+use xcm_builder::unique_instances::{derivatives::*, NonFungibleAsset};
+
+pub use pallet::*;
+
+/// The log target of this pallet.
+pub const LOG_TARGET: &'static str = "runtime::xnft";
+
+type DerivativeIdParamsOf<T, I> = <T as Config<I>>::DerivativeIdParams;
+
+type DerivativeIdOf<T, I> = <T as Config<I>>::DerivativeId;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	/// The in-code storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T, I = ()>(PhantomData<(T, I)>);
+
+	#[pallet::config]
+	/// The module configuration trait.
+	pub trait Config<I: 'static = ()>: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self, I>>
+			+ IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		type DerivativeIdParams: Member + Parameter + MaxEncodedLen;
+		type DerivativeId: Member + Parameter + MaxEncodedLen;
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn foreign_asset_to_derivative_id_params)]
+	pub type ForeignAssetToDerivativeIdParams<T: Config<I>, I: 'static = ()> =
```

Comment:
Why is it necessary to map `AssetId` to a `DerivativeId`. Is it possible to just use `AssetId`? As we do with `ForeignAssets` in asset hub.

Reply:
This is not strictly necessary. In fact, chains like AssetHub can use `AssetId` everywhere.

However, I introduced this optional pallet that can map `AssetId` to a different `DerivativeId` to support the use case of mapping foreign NFTs to the already existing NFT engine on a given chain. For instance, we at Unique want a unified interface for working with all kinds of NFTs on the chain, both ours and foreign ones. E.g., we want to use the existing methods for nesting and fractionalization.

This way, our clients can think about a single interface without worrying about the kind of token they are working with.

Reply:
Maybe the pallet's name should be refined to reflect its optionality.

Reply:
>I don't think it belongs in this repository then

Providing support for derivatives of a general form seems like a good thing. Also, the pallet is minimal, so maintaining it will be easy.